### PR TITLE
feat(serverless): add more technical details and troubleshooting section on custom domains

### PR DESCRIPTION
**Pull request by norbjd**
### Your checklist for this pull request

- [X] Name your pull request according to the [contribution guidelines](https://github.com/scaleway/docs-content/blob/main/docs/CONTRIBUTING.md).

### Description

I have added more technical details and troubleshooting stuff to the custom domains documentation.

**Don't merge until**:

- [x] markdown rendering and format is OK (I haven't checked how this is rendered)
- [x] @cyclimse and @thomas-tacquet have approved for the technical/product side
- [x] wording/english is OK (@bene2k1, @SamyOubouaziz is not available)
- [x] this doc is copied to "serverless functions" (same doc really)

### CODE REVIEW

File: serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx
Comment (bene2k1) [inline serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx:58]:
```suggestion
1. Configure the custom domain on our gateways, so they know that they should handle traffic for that custom domain.
```

File: serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx
Comment (bene2k1) [inline serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx:75]:
```suggestion
| DNS record is not propagated yet. | This can be the case if the custom domain is created immediately after the CNAME is configured on your DNS provider side. | As a rule of thumb, wait a few minutes after creating the CNAME on your DNS provider, or at least wait until the DNS record is propagated on main resolvers (`1.1.1.1`, `8.8.8.8`). |
```

File: serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx
Comment (bene2k1) [inline serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx:100]:
```suggestion
- downtime is not acceptable. Unfortunately, this is not possible right now out-of-the-box. To be able to answer to requests to `website.mydomain.com`, our infrastructure must know it, so custom domain has to be configured; but, for the custom domain to be configured, the DNS record must point to the Serverless Container endpoint, leaving us with a chicken-and-egg problem. To handle such scenarios, a third-party tool (like Cloudflare) can be configured to serve a cached version of your website while the domain is being reconfigured (for example with `stale-if-error` option). If you choose this solution, do not forget to disable cache on routes starting with `/.well-known/acme-challenge` to avoid issues described in the "Technical details and troubleshooting" section.
```

File: serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx
Comment (bene2k1) [inline serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx:41]:
```suggestion
    - running `dig`, `drill` or `nslookup` commands on your domain using common DNS resolvers `1.1.1.1` and `8.8.8.8`: `dig <your_custom_domain> @1.1.1.1`, `dig <your_custom_domain> @8.8.8.8`
    - or by using the DNS check propagation tool of your choice (some are available online)
```

File: serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx
Comment (thomas-tacquet) [inline serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx:88]:
```suggestion
- an `A` record is configured on `website.mydomain.com` and points to `51.15.x.x`
```

File: serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx
Comment (thomas-tacquet) [inline serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx:90]:
```suggestion
- you also have a running version of your website hosted as a Serverless Container (only accessible using `example-website.functions.fnc.fr-par.scw.cloud`), and now you want your users to access this version from `http://website.mydomain.com`
```

File: serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx
Comment (thomas-tacquet) [inline serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx:95]:
```suggestion
- after: `website.mydomain.com` is a CNAME record to `example-website.functions.fnc.fr-par.scw.cloud`
```

File: serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx
Comment (cyclimse) [inline serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx:41]:
```suggestion
  - or by using a DNS check propagation tool of your choice (some are available online)
```

File: serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx
Comment (thomas-tacquet) [inline serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx:97]:
```suggestion
By doing this, clients that already have the `website.mydomain.com` DNS record cached locally will continue to hit `51.15.x.x`, until the TTL expires. New clients (or those whose cache has expired) will start to hit `example-website.functions.fnc.fr-par.scw.cloud`. **However**, until the custom domain is not configured on the Serverless Container, these requests will end up in 404, because `website.mydomain.com` is not (yet) known in our infrastructure. Depending on your downtime tolerance (clients receiving 404) for a few minutes, there are 2 cases:
```

File: serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx
Comment (cyclimse) [inline serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx:58]: quibble: simplify wording

```suggestion
1. Configure the custom domain on our gateways, so that they can handle traffic for that custom domain.
```

File: serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx
Comment (cyclimse) [inline serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx:77]: nit: same, some wording simplification

```suggestion
| Negative DNS cache is stale | If the initial check fails (DNS record is not propagated yet), and the negative TTL configured on your DNS provider side is high, the negative TTL will prevent subsequent checks from querying the nameserver again, until it expires. Depending on the negative TTL configured, this can take more or less time. | Either reduce the negative TTL in your DNS provider configuration, or just wait until you are sure DNS record is propagated. If you already tried to add the custom domain and faced this issue, you likely have to wait for the negative TTL to expire before making another attempt (so the cache can also expire on our side). |
```

File: serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx
Comment (cyclimse) [inline serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx:99]: nit: same some wording

```suggestion
- downtime is acceptable or can be planned (e.g. during the night or when there is less traffic on your website). In that case, after creating the CNAME record, and once you are sure DNS is propagated everywhere, create the custom domain on the Serverless Container. Once ready, requests to `http://website.mydomain.com` will hit your Serverless Container (no more 404).
```

**Review by thomas-tacquet — APPROVED** [review]

File: serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx
Comment (cyclimse) [inline serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx:100]: nit: again just some wording, feel free to ignore

```suggestion
- downtime is not acceptable. Unfortunately, this is not possible right now out-of-the-box. To serve requests from `website.mydomain.com`, our infrastructure must know it, so a custom domain has to be configured. However, for the custom domain to be configured, the DNS record must point to the Serverless Container endpoint, resulting in a chicken-and-egg problem. To handle such scenarios, a third-party tool (like Cloudflare) can be configured to serve a cached version of your website while the domain is being reconfigured (for example with `stale-if-error` option). If you choose this solution, be sure to disable caching on routes starting with `/.well-known/acme-challenge` to avoid issues described in the "Technical details and troubleshooting" section.
```

**Review by cyclimse — APPROVED** [review]

**Review by Slashgear — APPROVED** [review]
Otherwise, this looks good to me ! 🆗

File: serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx
Comment (cyclimse) [inline serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx:74]: note: Dumb remark, but I feel like the `How to` is a rewording of the `Issue` column. Could we link to the doc/pic on what the CNAME record should look like?

File: serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx
Comment (cyclimse) [inline serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx:80]: nit: I feel like this title can be simplified, to me `already serving another endpoint` sounds strange. 

I don't think my suggestion is much better but maybe something like this:

```suggestion
## Migrating an existing website
```

Comment (norbjd) [reply serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx:80]: ah yes, way simpler, thanks!

File: serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx
Comment (norbjd) [reply serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx:39]: Ah yes you're right, DNS propagation is a misnomer. I'll check how I can replace "propagation" by something clearer (and more accurate).

File: serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx
Comment (norbjd) [reply serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx:74]: Yep, I'm not sure pictures are great in the table because they are too small, so I'll just add a text:

```suggestion
| DNS record is not correctly configured. | If the DNS record (mostly CNAME) is misconfigured on your DNS provider side, and the custom domain to the Serverless Container default endpoint, we will not be able to configure the custom domain. | Ensure that on your DNS provider, a CNAME record links your custom domain to the Serverless Container. To test, you can run the following command: `dig <your_custom_domain>` (or `nslookup`), and make sure that the value returned is a `CNAME` record to `<your_function>.functions.fnc.<region>.scw.cloud`. |
```

File: serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx
Comment (norbjd) [reply serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx]: I've replaced with "available" as it conveys the same meaning but it's not misleading at least. Couldn't find a better term :) https://github.com/scaleway/docs-content/pull/3593/commits/686127441a6a98df9f92b190bdd035376588d678

**Comment by bene2k1** [comment]
Need last review by @bene2k1 
Please do not merge

File: serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx
Comment (ldecarvalho-doc) [inline serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx:45]: to me this is not a step

File: serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx
Comment (ldecarvalho-doc) [inline serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx:46]:
```suggestion
7.  Connect to your container through that custom domain once the custom domain status is `ready`, you can
```

File: serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx
Comment (ldecarvalho-doc) [inline serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx:60]: maybe put the note in a message box

**Review by jsudraud — APPROVED** [review]

File: serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx
Comment (ldecarvalho-doc) [inline serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx:93]:
```suggestion
- when a client accesses `http://website.mydomain.com`, the request hits your Instance IP `51.15.x.x:80`
```

File: serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx
Comment (jcirinosclwy) [inline serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx:41]:
```suggestion
    - or using a DNS check tool of your choice (some are available online)
```

File: serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx
Comment (jcirinosclwy) [inline serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx:59]:
```suggestion
2. Ensure that the custom domain points to one of the Serverless Container existing endpoints: either the default one `....functions.fnc.<region>.scw.cloud`, or an existing custom domain.
```

File: serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx
Comment (jcirinosclwy) [inline serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx:90]:
```suggestion
- your website is hosted on a Scaleway Instance, with the IP `51.15.x.x`
```

**Review by jcirinosclwy — APPROVED** [review]

File: serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx
Comment (bene2k1) [inline serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx:45]:
```suggestion
    <Message type="note">
      The console runs checks on the CNAME and TLS configuration of the custom domain name. See technical details below for more details.
     </Message>
```

Comment (norbjd) [reply serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx:45]: Yes I agree, but since it was here before I've left it. Though this is misleading because it's not the console actually running the checks. I'll replace it with:

```suggestion
6. Wait while custom domain is being configured. See technical details below for more details.
```

Comment (norbjd) [reply serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx:45]: Well, more like in a note as @bene2k1 suggests below:

```suggestion
    <Message type="note">
      Before being `ready`, some checks are performed. See technical details below for more details.
    </Message>
```

File: serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx
Comment (norbjd) [reply serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx:60]:
```suggestion
3. If step 2 is OK, generate a dedicated TLS certificate for that custom domain so it can answer to HTTPS requests.

    <Message type="note">
      During that step, the custom domain might be available in plain HTTP. But, it will be accessible in HTTPS only once this step is over.
    </Message>
```

File: serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx
Comment (norbjd) [reply serverless/containers/how-to/add-a-custom-domain-to-a-container.mdx]: I've changed here: https://github.com/scaleway/docs-content/pull/3593/commits/428b80ba387f5ca600679096bdd5fd27a0d9de88

LMK what you think :+1:

**Comment by norbjd** [comment]
I've copied the "containers" doc to "functions" because it's really the same. I only made some adjustments to make it more coherent with functions (e.g. talk about web application and not website in the last section, as it does not really make sense to host a website on functions).

**Pull request merged**
